cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaDistributionGoal.java

```diff
@@ -377,13 +376,18 @@ private boolean rebalanceByMovingReplicasIn(Broker broker,
                                               Set<Goal> optimizedGoals,
                                               Set<String> excludedTopics) {
     PriorityQueue<Broker> eligibleBrokers = new PriorityQueue<>((b1, b2) -> {
-      int result = Double.compare(b2.replicas().size(), b1.replicas().size());
-      return result == 0 ? Integer.compare(b1.id(), b2.id()) : result;
+      // Brokers are sorted by (1) current offline replica count then (2) all replica count then (3) broker id.
+      int resultByOfflineReplicas = Integer.compare(b2.currentOfflineReplicas().size(), b1.currentOfflineReplicas().size());
+      if (resultByOfflineReplicas == 0) {
+        int resultByAllReplicas = Integer.compare(b2.replicas().size(), b1.replicas().size());
+        return resultByAllReplicas == 0 ? Integer.compare(b1.id(), b2.id()) : resultByAllReplicas;
+      }
+      return resultByOfflineReplicas;
     });
 
-    for (Broker healthyBroker : clusterModel.healthyBrokers()) {
-      if (healthyBroker.replicas().size() > _balanceLowerLimit) {
-        eligibleBrokers.add(healthyBroker);
+    for (Broker aliveBroker : clusterModel.aliveBrokers()) {
+      if (aliveBroker.replicas().size() > _balanceLowerLimit || !aliveBroker.currentOfflineReplicas().isEmpty()) {
```

Comment:
I think it is even stronger than that: if `_fixOfflineReplicasOnly ` is true, then add all other brokers, otherwise add all other brokers with replica count more than the limit or has offline replica.